### PR TITLE
getClassNames: results are memoized for better performance

### PR DESCRIPTION
**Pull request by dzearing**
#### Pull request checklist

- [X] Include a change request file using `$ npm run change`

#### Description of changes

The `classNamesFunction` factory returns a `getClassNames` function which will return classnames and mix things together properly. The original intent of the factory was to return a memoized function which would avoid doing extra work. This change fixes that intent to finally memoize responses from getStyles calls, resulting in much improved perf.

In the worst case scenario, the input comes in a constantly mutating format, such as an object. In this case, we'd be caching classnames unnecessarily and losing the benefit of memoization. This should be an edge case, but it should be avoided to add objects to any *StyleProps.

#### Focus areas to test

Things still render.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8761)

### CODE REVIEW

**Comment by KevinTCoughlin** [comment]
Woohoo! Will test this against OneDrive / SharePoint and more... Will report back.

cc: @aditima

**Comment by dzearing** [comment]
Working on the tests.

File: packages/utilities/src/classNamesFunction.test.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    Stylesheet.getInstance().setConfig({
+      onInsertRule: undefined
+    });
+  });
+
+  it('can cache rules', () => {
+    let styleFunctionCalled = false;
+
+    const getClassNames = classNamesFunction();
+    const getStyles = (props: { a: number }) => {
```
Comment (cliffkoh) [inline packages/utilities/src/classNamesFunction.test.ts]: should really just do `jest.fn.mockImplementation((props: ... => { ...})` here for getStyles, then you can get rid of styleFunctionCalled and do the far more idomatic `expect(getStyles).toHaveBeenCalledTimes(1)` below in line 33 and 36 (and in turn, `toHaveBeenCalledTimes(2)` in line 38 and 41)

File: packages/utilities/src/classNamesFunction.ts [hunk 2]
```diff
@@ -1,32 +1,73 @@
...
     styleFunctionOrObject: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
     styleProps: TStyleProps = {} as TStyleProps
   ): IProcessedStyleSet<TStyleSet> => {
-    // styleSet might be undefined if styleFunctionOrObject is undefined, but getStyles should never
-    // ordinarily be undefined (it would hardly make any sense).
-    // However, because we usually use `props.styles` as the argument to an invocation of this method, and
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
```
Comment (cliffkoh) [inline packages/utilities/src/classNamesFunction.ts]: shouldn't the template here be `Map<string, Map<string, any>>` at least, given the logic from line 33 to 47?

File: packages/utilities/src/classNamesFunction.ts [hunk 3]
```diff
@@ -1,32 +1,73 @@
...
+ * Note that the props you pass in on every render should be in the same order and
+ * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
+ * these will cause extra recalcs to occur.
  */
 export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(): (
   getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
   styleProps?: TStyleProps
 ) => IProcessedStyleSet<TStyleSet> {
-  // TODO: memoize.
+  let map: Map<any, any> = new Map<any, any>();
```
Comment (cliffkoh) [inline packages/utilities/src/classNamesFunction.ts]: should have comment explaining what this is a map of.. (what do the keys represent, what do values represent)

Comment (cliffkoh) [reply packages/utilities/src/classNamesFunction.ts]: Your scheme here is actually pretty cool but SERIOUSLY could use comments :) It also looks like the type can be better than `any, any`...

File: packages/utilities/src/classNamesFunction.ts (hunk 2 above)
Comment (cliffkoh) [reply packages/utilities/src/classNamesFunction.ts]: In TS 3, we can further refine to `Map<string, Map<string, unknown>>`

File: packages/utilities/src/classNamesFunction.ts [hunk 4]
```diff
@@ -1,32 +1,73 @@
...
+        if (!current.has(propValue)) {
+          current = current.set(propValue, new Map<any, any>());
+        }
+        current = current.get(propValue);
+      }
     }
 
-    const styleSet =
-      styleFunctionOrObject && (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);
+    if (!(current as any)[RetVal]) {
```
Comment (cliffkoh) [inline packages/utilities/src/classNamesFunction.ts]: I don't understand why do current[RetVal].. why not do `!current.has(RetVal)`? Am I missing something here

File: packages/utilities/src/classNamesFunction.ts [hunk 5]
```diff
@@ -1,32 +1,73 @@
...
+      if (styleFunctionOrObject === undefined) {
+        (current as any)[RetVal] = {} as IProcessedStyleSet<TStyleSet>;
+      } else {
+        const styleSet =
+          styleFunctionOrObject &&
+          (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);
+
+        (current as any)[RetVal] = mergeStyleSets(styleSet as TStyleSet);
```
Comment (cliffkoh) [inline packages/utilities/src/classNamesFunction.ts]: why not `current.set(RetVal, mergeStyleSets(styleSet as TStyleSet))` instead?

File: packages/utilities/src/classNamesFunction.ts [hunk 6]
```diff
@@ -1,32 +1,73 @@
...
+        resultCount++;
+      }
+    }
+
+    if (resultCount > MAX_CACHE_COUNT) {
+      if (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'production') {
```
Comment (cliffkoh) [inline packages/utilities/src/classNamesFunction.ts]: do you really mean !== production?

File: packages/utilities/src/classNamesFunction.ts [hunk 7]
```diff
@@ -1,32 +1,73 @@
...
+        throw new Error('Styles are being recalculated far too frequently.');
+      }
+      map.clear();
+    }
 
-    return mergeStyleSets(styleSet as TStyleSet);
+    return (current as any)[RetVal] as IProcessedStyleSet<TStyleSet>;
```
Comment (cliffkoh) [inline packages/utilities/src/classNamesFunction.ts]: why not current.get(RetVal)?

File: common/changes/@uifabric/utilities/getClassNames-perf_2019-04-18-05-32.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@uifabric/utilities",
+      "comment": "The `getClassNames` function now returns memoized classnames, resultng in improved performance in high repeat scenarios (like lists rendering checks.)",
```
Comment (khmakoto) [inline common/changes/@uifabric/utilities/getClassNames-perf_2019-04-18-05-32.json]: nit: resulting

**Comment by KevinTCoughlin** [comment]
Not sure we should automerge quite yet. I'm going to take a look at how this works out in a few partner scenarios and locally w.r.t perf. Will reply shortly.

File: packages/utilities/src/classNamesFunction.test.ts (hunk 1 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.test.ts]: thanks, will try it out!

File: packages/utilities/src/classNamesFunction.ts (hunk 2 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: They key is the value of the argument, which I don't know what it is. The value is either a string or another Map<any, string | Map<any, ...etc> > recursive. Not sure how to represent.

File: packages/utilities/src/classNamesFunction.ts (hunk 3 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: will do for sure, thanks.

File: packages/utilities/src/classNamesFunction.ts (hunk 6 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: oh great catch.

File: packages/utilities/src/classNamesFunction.ts (hunk 2 above)
Comment (cliffkoh) [reply packages/utilities/src/classNamesFunction.ts]: Oh you're right about the key. Sorry late night when I was reviewing. 🌙

Comment (cliffkoh) [reply packages/utilities/src/classNamesFunction.ts]: For Recursive, you can do

```
interface RecursiveMap extends Map<any, string | RecursiveMap> {    
}
```

then reference this everywhere you're using `Map<any, any>`.

**Comment by cliffkoh** [comment]
@msft-github-bot make sure @KevinTCoughlin approves.

**Comment by dzearing** [comment]
Doing lots of testing before the merge. I think the snapshots are raising issues here.

**Comment by KevinTCoughlin** [comment]
FYI - If folks ever need to quickly disable tests and still build, I found the following diff to be the easiest path:

```diff
diff --git a/scripts/jest.config.js b/scripts/jest.config.js
index 7061c6003..034f957aa 100644
--- a/scripts/jest.config.js
+++ b/scripts/jest.config.js
@@ -2,7 +2,7 @@ const config = {
   moduleNameMapper: {
     'office-ui-fabric-react/lib/(.*)$': '<rootDir>/src/$1'
   },
-  testRegex: '(/__tests__/.*|\\.(test|spec))\\.(js)$',
+  testRegex: '(/__tests__/.*|\\.(foo))\\.(js)$',
   transform: {},
   testEnvironment: 'node'
 };
diff --git a/scripts/jest/jest-resources.js b/scripts/jest/jest-resources.js
index dd31ae12e..569d3eeba 100644
--- a/scripts/jest/jest-resources.js
+++ b/scripts/jest/jest-resources.js
@@ -5,11 +5,10 @@ const resolve = require('resolve');
 module.exports = {
   createRawConfig: () => ({
     rootDir: 'lib',
-    testRegex: '(/__tests__/.*|\\.(test|spec))\\.js$'
+    testRegex: '(/__tests__/.*|\\.(foo))\\.js$'
   }),
   createConfig: customConfig =>
-    merge(
-      {
+    merge({
         moduleNameMapper: {
           'ts-jest': resolve.sync('ts-jest'),
           '\\.(scss)$': path.resolve(__dirname, 'jest-style-mock.js'),
@@ -22,7 +21,7 @@ module.exports = {
 
         reporters: [path.resolve(__dirname, './jest-reporter.js')],
 
-        testRegex: '(/__tests__/.*|\\.(test|spec))\\.(ts|tsx)$',
+        testRegex: '(/__tests__/.*|\\.(foo))\\.(ts|tsx)$',
         moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json'],
 
         setupFiles: [path.resolve(__dirname, 'jest-setup.js')],
```

**Comment by KevinTCoughlin** [comment]
### Preliminary results

The below results were gathered in Chrome Canary `75.0.3769.0` with the following:

- URL used: http://localhost:4322/#/tests/detailslistbasicexample
- Example used: DetailsList.Basic.Example.tsx
- Virtualization: **Disabled** to emphasize per row render delta
- Count of items rendered: 1000
- React Production: TRUE
- Total time gathered from profiler's `_onLoad` flame (initial render of example)

Each run, I refreshed via the Chrome profiler's "refresh" button and captured initial render time in Excel.

  | Run 1 (ms) | Run 2 (ms) | Run 3 (ms) | Run 4 (ms) | Run 5 (ms) | Average (ms) | Time Per Row (ms)
-- | -- | -- | -- | -- | -- | -- | --
Before Cache | 1160 | 1090 | 1130 | 1430 | 1180 | 1198 | 1.198
After Cache | 694.85 | 727.71 | 744.11 | 700.39 | 743.70 | 722.152 | 0.722152

File: packages/utilities/src/classNamesFunction.ts (hunk 2 above)
Comment (KevinTCoughlin) [reply packages/utilities/src/classNamesFunction.ts]: That's cool!

File: packages/utilities/src/classNamesFunction.ts (hunk 4 above)
Comment (KevinTCoughlin) [reply packages/utilities/src/classNamesFunction.ts]: Not sure if it is the same in TS/JS, but isn't it normally recommended to simply get and check `undefined | null` vs. `has` re: performance? Thought that was the case in Java / C#... would have to measure.

**Comment by cliffkoh** [comment]
@KevinTCoughlin you can use `rush build --min` to avoid running tests to generate a build.

**Comment by KevinTCoughlin** [comment]
Here are the results from yesterday's test against a document library:


Iteration | 1 | 2 | 3 | 4 | 5 | Average (ms)
-- | -- | -- | -- | -- | -- | --
Before (ms) | 6049 | 6626 | 6700 | 5651 | 5865 | 6178.2
After (ms) | 5988 | 5908 | 5992 | 5912.00 | 5995 | 5959
  |   |   |   |   | Delta | 219.2

**Comment by dzearing** [comment]
Quite update, I have a number of updates locally, but found a number of cases where this won't work:

* if you pass in a styles function into a component (this really can't be memoized)
* If a child component is rendered within a parent which uses subcomponent styles

**Comment by dzearing** [comment]
I do not understand why this has failed to build. It builds locally. I went to another machine and pulled the branch down and built without issue. Re-enqueuing the build.

File: packages/utilities/src/classNamesFunction.ts (hunk 7 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: I will add a comment here, but because RetVal is an attached property; not a key in the hash. :)

File: packages/utilities/src/classNamesFunction.ts [hunk 9]
```diff
@@ -1,33 +1,138 @@
...
+      options.disableCaching = true;
+
+      // Note: this code is great for debugging problems with styles being recaculated, but I'm commenting it out
+      // to avoid confusing consumers.
+      
+      // if (process.env.NODE_ENV !== 'production') {
+      //  console.log('Styles are being recalculated far too frequently. Something is mutating the class over and over.');
+      //  // tslint:disable-next-line:no-console
+      //  console.trace();
+      }
```
Comment (JasonGore) [inline packages/utilities/src/classNamesFunction.ts]: Hmm maybe this could be an `option` only used when not in `production`?

File: packages/utilities/src/classNamesFunction.ts [hunk 10]
```diff
@@ -1,33 +1,138 @@
...
-    // ordinarily be undefined (it would hardly make any sense).
-    // However, because we usually use `props.styles` as the argument to an invocation of this method, and
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
+    const disableCaching = options.disableCaching;
+
+    if (!options.disableCaching) {
+      current = _traverseMap(map, styleFunctionOrObject as any);
```
Comment (JasonGore) [inline packages/utilities/src/classNamesFunction.ts]: What happens here if `styleFunctionOrObject` is a function? It seems like `_traverseMap` will traverse keys of the function, but is that right? Shouldn't it traverse the style function output?

File: packages/utilities/src/classNamesFunction.ts [hunk 11]
```diff
@@ -1,33 +1,138 @@
...
+      current = _traverseMap(current, styleProps);
```
Comment (JasonGore) [inline packages/utilities/src/classNamesFunction.ts]: I'm curious if it's true that objects are allowed in `styleFunctionOrObject` but could be a red flag for `styleProps`? (This ties to my other comment where I'm not sure if `theme` is the reason why multiple components are using `disableCaching`.) If that's true, I'm wondering if we could generate an optional warning if objects are detected in the `styleProps` case as a possible indicator that `disableCaching` should be used.

File: packages/utilities/src/classNamesFunction.ts (hunk 10 above)
Comment (JasonGore) [reply packages/utilities/src/classNamesFunction.ts]: Ahh, I think I see now. You can use the function as a key since you're evaluating it against the `styleProps` below. If either change, it will traverse the map appropriately.

File: packages/office-ui-fabric-react/src/components/Button/examples/Button.Command.Example.tsx [hunk 12]
```diff
@@ -4,11 +4,11 @@ import { Label } from 'office-ui-fabric-react/lib/Label';
 import { css, classNamesFunction } from 'office-ui-fabric-react/lib/Utilities';
 import { getStyles, IButtonBasicExampleStyleProps, IButtonBasicExampleStyles } from './Button.Basic.Example.styles';
 
+const getClassNames = classNamesFunction<IButtonBasicExampleStyleProps, IButtonBasicExampleStyles>();
```
[the remark this answers is not in the document]
Comment (JasonGore) [reply packages/office-ui-fabric-react/src/components/Button/examples/Button.Command.Example.tsx]: Converting `getStyles` to objects also means that classNames don't have to be generated in any of these renders, which I've also fixed and moved out.

File: packages/utilities/src/classNamesFunction.ts (hunk 2 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: I think I'm going to pass on being clever on this one.

File: packages/utilities/src/classNamesFunction.ts (hunk 4 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: I added comments, but the `[RetVal]` is an attached property, not a key in the Set. It represents the return value for the branch in the graph.

File: packages/utilities/src/classNamesFunction.ts (hunk 5 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: Same as above.

File: packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.base.tsx [hunk 13]
```diff
@@ -42,8 +42,12 @@ import { IContextualMenuItemStyleProps, IContextualMenuItemStyles } from './Cont
...
 import { getItemStyles } from './ContextualMenu.classNames';
 
-const getClassNames = classNamesFunction<IContextualMenuStyleProps, IContextualMenuStyles>();
-const getContextualMenuItemClassNames = classNamesFunction<IContextualMenuItemStyleProps, IContextualMenuItemStyles>();
+const getClassNames = classNamesFunction<IContextualMenuStyleProps, IContextualMenuStyles>({
+  disableCaching: true
+});
+const getContextualMenuItemClassNames = classNamesFunction<IContextualMenuItemStyleProps, IContextualMenuItemStyles>({
+  disableCaching: true
+});
```
[the remark this answers is not in the document]
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.base.tsx]: As mentioned, I've simply disabled caching for things which I know will mutate frequently (due to closured subcomponent usage which I can't really make better without a more thorough intrusive fix.)

File: packages/utilities/src/classNamesFunction.ts (hunk 11 above)
Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: We could but in the worse case scenario we auto disable caching for these scenarios.

Comment (dzearing) [reply packages/utilities/src/classNamesFunction.ts]: I don't really want to confuse partners by having the debug experience throwing warnings, unless they're actionable.
